### ISSUE TITLE
Save past requests to API in the database

### ISSUE THREAD

**Issue by Frollamma** [parmigggiana/attack_scripts#7]
At the moment, when we launch for the first time an exploit it takes as argument just the ip, it performs a request to the API, then it parses to get the relevant vulnerable stuff for that ip. This is not efficient because for each ip and for each service we're performing an API request, we actually need to perform that request only once, save the result and pass it to to the exploit function.

**Comment by parmigggiana** [comment]
This is a good idea in the context of the Cyberchallenge but doing the call in the program and then passing the returned value to the exploits is a big loss in generality for other CTFs. We should consider just doing this in a module and then importing it in the other exploits.

**Comment by Frollamma** [comment]
How do you want to implement that using a module exactly?

My idea is that we may generalize the current approach so that it fits our needs as well, without using a module, because I can't think of any "non tricky" way of doing this for now. I think that the most general way of doing this is the following:
- Impose a "target centric" model to exploits: In the competition we had a list of ips and for each ip we had some emails, for each ip we had to connect to the service and exploits the accounts relative to the emails related to that ip. If we define as "target" as "sufficient data we need to start the exploit", given an ip our target to get all the flags for that ip was the couple $(ip, [email_1, email_2, \dots, email_n])$, so all the "targets" we needed to get all the possible flags can be seen as the set $\{(ip_1, [email_{1 1}, email_{1 2}, \dots, email_{1 n_1}]), \dots, (ip_m, [email_{m 1}, email_{m 2}, \dots, email_{m n_m}])\}$ (in python you can think of it as a list). Notice that, when possible, we want to pass also the "minimal sufficient data we need to start the exploit", like in the previous example. Ok, so "imposing a target centric model to exploits" means that every exploit takes as input a set of targets, defined like in the previous definition. Now you might say "but this definition is pretty vague", this is the whole point! We define exactly what a "target" is case by case, depending on our needs. This way we keep the code extremely general, the only thing we're asking here is to pass to the exploits a set of targets (you can also pass a set with a single element of course...). Now we need a way to define in practice what a target is, we'll see it the next point in this list. This way when the concept of "target" is defined the exploits know the format of the input they receive.
- Define a `target_retriever` function (you can find a better name): Milkman has to run all the exploits ok, but Milkman has also to pass some data to them, at the moment I guess it just passes ips, now we have a "target centric model", this means that we have to pass targets to exploits, not ips. So in order to get the date milkman runs a function `target_retriever` that retrieves all the targets, for example this function can read a file and parse it to return a set of targets or this function can make a request to a server (like in our case) and parse it to return a set a of targets. Remember that this function has always to return a set of targets. To make this even more cool you could, in python, add a class named "Target", so that is we have an object `target` we can do things like `target.ip` or `target.emails` depending on the specific CTF

To summarize:
- User defines in the configuration the function `target_retriever` (or maybe you should use a class "ExploitHandler" and define `target_retriever` as mathod or something, I don't know, it looks more professional)
- Milkman gets targets using `target_retriever`
- exploits get as input targets as `target_retriever` outputs them

**Comment by parmigggiana** [comment]
There's more than one problem with this approach:
1. while it's technically possible to generalize the arguments using kwargs, how do you want to generalize what arguments you pass to each thread? Every exploit may need different extra info (or none at all); you would need to statically write in the 'deployer' (the function launching all the exploits) what to give to each process, or at best pass to everyone a dictionary with all the available info. 
2. You say we should retrieve this dictionary with a user defined function, but where would you put that in the code? I don't think it's a reasonable expectation for users to be modifying the inner code of the app, that's why I said it's just much simpler to do it in a module and have the exploits that need that dictionary call the module. That would be the most general and simple approach, since we can't assume every ctf presents data the same way (or even at all)

In the end, any information an exploit requires to run besides the target ip is exploit-specific and should be handled by who writes the exploit, not the library.

**Comment by Frollamma** [comment]
We should update this thread with what we said in the recent days. @parmigggiana

**Comment by Frollamma** [comment]
@parmigggiana proposed to let the user create a class that handles the API request to get the targets, then it parses it in order to be used by the exploit. The exploit won't accept a "target object" as I described it before, it will get the arguments it needs to run the exploit. The advantage of this technique is that if the same target has got multiple vulnerabilities the same class can be reused for different exploit, without having to rewrite the parsing part in new exploits. The disadvantage is that if an exploiter created an exploit, they probably didn't use a class to make the parsing, so there are some modifications that they have to do before deploying the exploit. @parmigggiana tell me if that's what you thought. By the way, I think that a parsing function would be better, unless you want to abstract even more

**Comment by parmigggiana** [comment]
I have added the issue #15 which tackles a similar issue. I'm keeping this thread open for now, but I feel we should move the discussion there. The solution I'm proposing would make it possible to solve this one as well. If we have a dictionary-like structure the exploit could reference all the target info for his designated target (that would still be the target ip, used as key to look what the exploit actually needs in the global target dictionary). After running the exploit, it can pop the used values from the dictionary. Of course when we build the struct we only append, without overwriting older calls. This way, we have process that only adds, and a process that only removes (after using)
